### ISSUE TITLE
If an output plugin stop retrieving elements from the queue, the stop sequence does not work

### ISSUE THREAD

**Issue by purbon** [elastic/logstash#3491]
If an output plugin stops retrieving elements from their queue, logstash is going to stall as already known, but if this happen during the shutdown sequence then logstash is not able to finish gratefully as expected. 

The bottleneck code can be seen for example from:

```
 def shutdown_filters
    @flusher_thread.kill
    @input_to_filter.push(LogStash::SHUTDOWN)
  end
```
https://github.com/elastic/logstash/blob/master/lib/logstash/pipeline.rb#L117-L120

or from:

```
 def shutdown_outputs
    # nothing, filters will do this
    @filter_to_output.push(LogStash::SHUTDOWN)
  end
```
https://github.com/elastic/logstash/blob/master/lib/logstash/pipeline.rb#L126-L129

imagine the situation where the output plugin is into an infinite loop of retry, even if an error is happening, then the queues are going to fill up and the ```LogStash::SHUTDOWN``` event is never going to be in, until the back pressure is relieved.

For now, until the semantic of the pipeline is change to handle this, output plugins should be responsible or not entering infinite loops.

One example of this can be seen at https://github.com/logstash-plugins/logstash-output-redis/pull/9 where in case of redis being offline, the output enters an infinite loops of retries, never going to consume any new event, so stalling everything.

**Comment by purbon** [comment]
@jsvd I think you might be interested on this as of your pipeline research.

**Comment by andrewvc** [comment]
Good catch @purbon

**Comment by colinsurprenant** [comment]
@purbon this is a known issue, was first reported in a related situation in #2130 and I also created a ticket specifically for this in #3451 I thin we can close it as a duplicate.

**Comment by purbon** [comment]
Closing this and adding a note about infinite retries causing this issue at #3451, more details https://github.com/elastic/logstash/issues/3451#issuecomment-115193093.

@colinsurprenant thanks for pointing this as duplicate, I was not aware this was a known problem. I guess we should aim to review the most common outputs to see if we can do anything to prevent known causes of this issue. For example as I proposed for the redis output, does this makes sense to you?
